java.base/src/main/java/java/net/NetworkInterface$_qbicc.java

```diff
@@ -0,0 +1,537 @@
+ *
+ * ------
+ * Copyright (c) 2001, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ * ------
+ *
+ * This file may contain additional modifications which are Copyright (c) Red Hat and other
+ * contributors.
+ */
+
+package java.net;
+
+import static org.qbicc.runtime.CNative.*;
+import static org.qbicc.runtime.bsd.Ifaddrs.*;
+import static org.qbicc.runtime.bsd.NetIf.*;
+import static org.qbicc.runtime.linux.NetIf.*;
+import static org.qbicc.runtime.posix.ArpaInet.*;
+import static org.qbicc.runtime.posix.Errno.*;
+import static org.qbicc.runtime.posix.SysIoctl.*;
+import static org.qbicc.runtime.posix.NetinetIn.*;
+import static org.qbicc.runtime.posix.NetIf.*;
+import static org.qbicc.runtime.posix.SysSocket.*;
+import static org.qbicc.runtime.posix.Unistd.*;
+import static org.qbicc.runtime.stdc.Errno.*;
+import static org.qbicc.runtime.stdc.Stddef.*;
+import static org.qbicc.runtime.stdc.Stdint.*;
+import static org.qbicc.runtime.stdc.Stdlib.*;
+import static org.qbicc.runtime.stdc.String.*;
+
+import org.qbicc.rt.annotation.Tracking;
+import org.qbicc.runtime.Build;
+
+@Tracking("src/java.base/unix/native/libnet/NetworkInterface.c")
+@Tracking("src/java.base/windows/native/libnet/NetworkInterface.c")
+class NetworkInterface$_qbicc {
+
+    static class netaddr {
+        struct_sockaddr_ptr addr;
+        struct_sockaddr_ptr brdcast;
+        c_short mask;
+        c_int family;
+        netaddr next;
+    }
+
+    static class netif {
+        char_ptr name;
+        c_int index;
+        boolean virtual;
+        netaddr addr;
+        netif childs;
+        netif next;
+    }
+
+    /*
+     * Frees the malloced struct_sockaddrs and char_ptrs for an interface list (including any attached addresses).
+     */
+    private static void freeif(netif ifs) {
```

Comment:
Do I understand correctly that we deliberately pass a structure by value here?

Reply:
`netif` and `netaddr` were C structs defined locally in OpenJDK's NetworkInterface.c.  When I ported the code to Java, I didn't try to define them as native structs, I made them Java classes instead.

Reply:
Ah, I completely missed that they are not structures. I'll have to re-review the patch now with that insight...

Reply:
I could make the structs extend `object`, then it would really parallel the C code, including the need to malloc/free instances of them since we're building up a dynamic linked structure.   If we went down that route, I'd probably also bring back the cleverness of doing combined malloc operations for both the struct and its contained sub-structs.